### PR TITLE
Feature Engineering

### PR DESCRIPTION
**Pull request by Hwailaluta**
My variables were: exterior1st, exterior2nd, roofmatl, lotarea, lotfrontage, masvnrarea, BsmtFinSF1, BsmtFinSF2, and masvnrtype.

### CODE REVIEW

**Review by srwight — COMMENTED** [review]
Waiting to approve until we hear back from Kyle about duplicated effort

File: hend.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    #Imputing missing numerical data with the mean and categorical data with the mode.
+    df['LotFrontage'].fillna(df['LotFrontage'].mean(), inplace=True)
+    df['MasVnrArea'].fillna(df['MasVnrArea'].mode(), inplace=True)
+
+    #Determing interquartile range 
+    Q1 = df['LotFrontage'].quantile(.25)
+    Q3 = df['LotFrontage'].quantile(.75)
+    IQR = Q3 - Q1
+
+    #I replaced outliers with the mean.
```
Comment (Hwailaluta) [inline hend.py]: *Clipped, not mean.

File: hend.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    for entry in df['MasVnrArea']:
+    if entry > (Q3 + 1.5 * IQR):
+        df['MasVnrArea'].replace(entry, df['MasVnrArea'].mean(), inplace = True)
+
+    #Dummy variables are given for categorical variables.
+    pd.get_dummies(df['Exterior1st'])
+    pd.get_dummies(df['RoofMatl'])
+    pd.get_dummies(df['MasVnrType'])
+
```
Comment (srwight) [inline hend.py]: get_dummies doesn't put them into the dataframe. It returns them. You'll need to capture them in a variable and add them back to your frame.

File: hend.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    df['LotFrontage'] = np.log(df['LotFrontage'] + 1)
+    df['MasVnrArea'] = np.log(df['MasVnrArea'] + 1)
+    df['LotArea'] = np.log(df['LotArea'] + 1)
+
+    #Due to the presence of 'TotalBsmtSF' (which is a sum of the below two variables) and sparseness of data, I drop these two.
+    #df.drop(['BsmtFinSF1'], axis = 1, inplace=True)
+    #df.drop(['BsmtFinSF1'], axis = 1, inplace=True)
+
+    features = df[['Exterior1st', 'RoofMatl', 'MasVnrType', 'MasVnrArea', 'LotFrontage', 'LotArea']]
+
```
Comment (srwight) [inline hend.py:21]: Your return df will have other columns, too, which will be the dummies you made up above.

File: hend.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    pd.get_dummies(df['MasVnrType'])
+
+    #Applying log+1 to each numerical category.
+    df['LotFrontage'] = np.log(df['LotFrontage'] + 1)
+    df['MasVnrArea'] = np.log(df['MasVnrArea'] + 1)
+    df['LotArea'] = np.log(df['LotArea'] + 1)
+
+    #Due to the presence of 'TotalBsmtSF' (which is a sum of the below two variables) and sparseness of data, I drop these two.
+    #df.drop(['BsmtFinSF1'], axis = 1, inplace=True)
+    #df.drop(['BsmtFinSF1'], axis = 1, inplace=True)
```
Comment (srwight) [inline hend.py]: Is this a copy/paste error? you have the same line twice, which will throw an error when the second line can't find the column.

File: hend.py [hunk 5]
```diff
@@ -30,17 +23,13 @@ def feature_extract(df):
     IQR = Q3 - Q1
 
     #I replaced outliers with the mean.
-    for entry in df['LotFrontage']:
-        if entry > (Q3 + 1.5 * IQR):
-            df['LotFrontage'].replace(entry, df['LotFrontage'].mean(), inplace = True)
+    df['LotFrontage'].clip(0,((Q3 + 1.5 * IQR))
```
Comment (srwight) [inline hend.py]: You have a hanging paren '(' after clip that isn't resolved. It looks like your goal is to clip the data to either zero or the 3rd quartile plus 1.5 IQR's, right? That looks good. But you might want to drop one of the '(' before Q3. I think that will solve your problem.

File: hend.py [hunk 6]
```diff
@@ -30,17 +23,13 @@ def feature_extract(df):
...
 
     Q1 = df['MasVnrArea'].quantile(.25)
     Q3 = df['MasVnrArea'].quantile(.75)
     IQR = Q3 - Q1
 
-    for entry in df['MasVnrArea']:
-    if entry > (Q3 + 1.5 * IQR):
-        df['MasVnrArea'].replace(entry, df['MasVnrArea'].mean(), inplace = True)
+    df['MasVnrArea'].clip(0,((Q3 + 1.5 * IQR))
```
Comment (srwight) [inline hend.py]: Same issue as above

File: hend.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    #df.drop(['BsmtFinSF1'], axis = 1, inplace=True)
+    #df.drop(['BsmtFinSF1'], axis = 1, inplace=True)
+
+    features = df[['Exterior1st', 'RoofMatl', 'MasVnrType', 'MasVnrArea', 'LotFrontage', 'LotArea']]
+
+    return features
+
+def main():
+    df=pd.read_csv('train.csv')
+    print(feature_extract(df))
```
Comment (srwight) [inline hend.py:27]: you still had 'engineer()' here. I swapped it out.

**Review by srwight — CHANGES_REQUESTED** [review]
See comments within - you have some syntax issues, and your feature variables are not getting captured.

**Review by srwight — APPROVED** [review]
Yup

**Review by disassembledd — APPROVED** [review]

**Pull request merged**